File: gali/sources/lutris_source.py
from sqlite3 import connect, Row

from gali.utils.locations import HOME
from gali.sources.source import Source
from gali.games.lutris_game import LutrisGame


class LutrisSource(Source):

    name: str = "Lutris"
    game_class: type[LutrisGame] = LutrisGame

    db_path: str = f"{HOME}/.local/share/lutris/pga.db"
    db_request: str = """
        SELECT
            name, slug, configpath, installed
        FROM
            'games'
        WHERE
            NOT hidden
            AND name IS NOT NULL
            AND slug IS NOT NULL
            AND configPath IS NOT NULL
        ;
    """

    def get_db_contents(self) -> list[Row]:
        connection = connect(self.db_path)
        cursor = connection.execute(self.db_request)
        rows = cursor.fetchall()
        connection.close()
        return rows

    def get_games(self, rows: list[Row]) -> tuple[LutrisGame]:
        games = []
        for row in rows:

            name = row[0]
            game_slug = row[1]
            config_path = row[2]
            is_installed = row[3]

            # Skip broken games
            if (
                name is None or
                game_slug is None or
                config_path is None
            ):
                continue

            game = self.game_class(
                name=name,
                game_slug=game_slug,
                config_path=config_path,
                is_installed=is_installed
            )
            games.append(game)

        return tuple(games)

    def scan(self) -> tuple[LutrisGame]:
        db_rows = self.get_db_contents()
        games = self.get_games(db_rows)
        return games

File: src/scanners/LutrisSource.js
const sqlite3 = require("sqlite3");
const sqlite = require("sqlite");
const fs = require("fs");

const Source = require("./Source.js");
const LutrisGame = require("../games/LutrisGame");

const USER_DIR = process.env["HOME"];

class LutrisSource extends Source {

	static name = "Lutris";
	static gameClass = LutrisGame;

	preferCache = false;

	dbPath = `${USER_DIR}/.local/share/lutris/pga.db`;
	bannerPath = `${USER_DIR}/.local/share/lutris/banners`;
	iconPath = `${USER_DIR}/.local/share/icons/hicolor/128x128/apps`;
	databaseRequest = `
		SELECT 
			name, slug, configpath, installed 
		FROM
			'games' 
		WHERE 
			NOT hidden
			AND name IS NOT NULL 
			AND slug IS NOT NULL
			AND configPath IS NOT NULL
		;
	`;

	constructor(preferCache = false) {
		super();
		this.preferCache = preferCache;
	}

	/**
	 * Optional step, add images to a game
	 * @param {LutrisGame} game - The game to get image for
	 */
	_getGameImages(game) {
		const images = {
			coverImage: `${this.bannerPath}/${game.gameSlug}.jpg`,
			iconImage: `${this.iconPath}/lutris_${game.gameSlug}.png`,
		};
		for (const [key, value] of Object.entries(images)) {
			const imageExists = fs.existsSync(value);
			if (imageExists) {
				game[key] = value;
			}
		}
	}

	/**
	 * Get all Lutris games
	 * @returns {LutrisGame[]} - An array of found games
	 */
	async scan() {

		// Open DB
		const db = await sqlite.open({
			filename: this.dbPath,
			driver: sqlite3.cached.Database
		});

		// Get games
		const games = [];
		const results = await db.all(this.databaseRequest);
		for (const row of results) {
			const game = new this.constructor.gameClass(
				row.slug,
				row.name,
				row.configpath,
				row.installed
			);
			this._getGameImages(game);
			games.push(game);
		}
		return games;

	}

}

module.exports = LutrisSource;